scripts/subspace_learning_lib.py

```diff
@@ -0,0 +1,243 @@
+
+        kernel_generator = lambda step_size, inverse_mass_matrix: jit(nuts.kernel(
+            potential, step_size, inverse_mass_matrix))
+
+        stan_key, key = split(rng_key)
+
+        final_state, (step_size, inverse_mass_matrix), _ = stan_warmup.run(
+            stan_key,
+            kernel_generator,
+            initial_state,
+            num_warmup)
+
+        nuts_kernel = kernel_generator(step_size, inverse_mass_matrix)
+
+        inference_key, key = split(key)
+        states = inference_loop(inference_key, nuts_kernel, final_state, num_samples)
+        return states.position
+
+    return nuts_sampler
+
+
+def build_sgd(train_step, dataset, batch_size):
+    dataloader = data_stream(dataset, batch_size)
+    train_step_partial = partial(train_step, dataloader=dataloader)
+
+    def sgd(_, num_samples, params):
+        iterations = jnp.arange(num_samples)
+        mass = jnp.zeros(params.shape)
+        velocity = jnp.zeros(params.shape)
+
+        @progress_bar_scan(num_samples)
+        def train_step_(*args, **kwargs):
+            return train_step_partial(*args, **kwargs)
+
+        (final_params, _, _), _ = scan(train_step_, (params, mass, velocity), iterations)
+        return final_params[None, ...]
+
+    return sgd
+
+
+def subspace_learning(key, num_samples, num_warmup, dataset, d, loglikelihood_fn, logprior_fn, params, sampler_init_fn,
+                      predict_fn, hyperparams):
+    initial_full_params, flat_to_pytree_fn = jax.flatten_util.ravel_pytree(params)
+    D = len(initial_full_params)
+    subspace_key, key = split(key)
+    projection_matrix = generate_random_basis(subspace_key, d, D)
+
+    def subspace_train_callback(params_subspace, dataset):
+        params_full = convert_params_from_subspace_to_full(params_subspace, projection_matrix, initial_full_params)
+        params_pytree = flat_to_pytree_fn(params_full)
+        epoch_accuracy, _ = partial(accuracy, predict_fn=predict_fn)(params_pytree, dataset)
+        return epoch_accuracy
+
+    def projected_loglikelihood(params_subspace, *data):
+        params_full = convert_params_from_subspace_to_full(params_subspace, projection_matrix, initial_full_params)
+        params_pytree = flat_to_pytree_fn(params_full)
+        return partial(loglikelihood_fn, predict_fn=predict_fn)(params_pytree, *data)
+
+    def projected_logprior(params_subspace):
+        params_full = convert_params_from_subspace_to_full(params_subspace, projection_matrix, initial_full_params)
+        return logprior_fn(params_full)
+
+    def projected_potential(params_subspace, batch):
+        X, y = batch
+        params_full = convert_params_from_subspace_to_full(params_subspace, projection_matrix, initial_full_params)
+        params_pytree = flat_to_pytree_fn(params_full)
+        batch_loglikelihood = vmap(loglikelihood_fn, in_axes=(None, 0, 0, None))(params_pytree, X, y, predict_fn)
+        return -jnp.sum(batch_loglikelihood) - projected_logprior(params_subspace)
```

Comment:
The potential should be defined as
```
U(theta) = -log_prior(theta) -N/B * sum_{i=1}^B loglik(i) 
```
where N is the total size of the dataset, and B is the minibatch size.